### PR TITLE
Fix code coverage test

### LINKED ISSUE #254: Unexpected behavior for coverage tests
The default `CODE_COVERAGE` option is off, still the test with coverage are created by default (tested on Summit with cmake 3.14.1)
```bash
$ ctest -N
Test project /gpfs/alpine/ast146/world-shared/build-cuda
  Test  #1: unit_tests:Just check everything
  Test  #2: unit_tests:Can create a vector-valued face-variable
  Test  #3: unit_tests:Add and Get is called
  Test  #4: unit_tests:reset is called
  Test  #5: unit_tests:when hasKey is called
  Test  #6: unit_tests:Physical constants
  Test  #7: unit_tests:Checking IndexShape indices
  Test  #8: unit_tests:Checking IndexShape cell counts
  Test  #9: unit_tests:par_for loops
  Test #10: unit_tests:nested par_for loops
  Test #11: unit_tests:Overlapping SpaceInstances
  Test #12: unit_tests:Built-in flags are registered
  Test #13: unit_tests:A Metadata flag is allocated
  Test #14: unit_tests:A Metadata struct is created
  Test #15: unit_tests:ParArrayND
  Test #16: unit_tests:ParArrayND with LayoutLeft
  Test #17: unit_tests:Time simple stencil operations
  Test #18: unit_tests:Check registry pressure
  Test #19: unit_tests:Check many arrays
  Test #20: unit_tests:Can pull variables from containers based on Metadata
  Test #21: unit_tests:Test required/desired checking from inputs
  Test #22: unit_tests:Parthenon Error Checking
  Test #23: performance_tests:Catch2 Container Iterator Performance
  Test #24: regression_test:output_hdf5
  Test #25: regression_coverage_test:output_hdf5
  Test #26: regression_test:calculate_pi
  Test #27: regression_coverage_test:calculate_pi
  Test #28: regression_test:advection_convergence
  Test #29: regression_coverage_test:advection_convergence
  Test #30: regression_test:advection_performance
  Test #31: regression_coverage_test:advection_performance
```

The behavior I expected would be that there are no coverage tests when `CODE_COVERAGE=OFF`.

The second unexpected behavior pertains to the labels.
Of the 31 tests above the "coverage" labels are assigned as follows:
```
$ ctest -N -L "coverage"
Test project /gpfs/alpine/ast146/world-shared/build-cuda
  Test  #1: unit_tests:Just check everything
  Test  #2: unit_tests:Can create a vector-valued face-variable
  Test  #3: unit_tests:Add and Get is called
  Test  #4: unit_tests:reset is called
  Test  #5: unit_tests:when hasKey is called
  Test  #7: unit_tests:Checking IndexShape indices
  Test  #8: unit_tests:Checking IndexShape cell counts
  Test  #9: unit_tests:par_for loops
  Test #10: unit_tests:nested par_for loops
  Test #12: unit_tests:Built-in flags are registered
  Test #13: unit_tests:A Metadata flag is allocated
  Test #14: unit_tests:A Metadata struct is created
  Test #15: unit_tests:ParArrayND
  Test #16: unit_tests:ParArrayND with LayoutLeft
  Test #20: unit_tests:Can pull variables from containers based on Metadata
  Test #21: unit_tests:Test required/desired checking from inputs
  Test #22: unit_tests:Parthenon Error Checking
  Test #25: regression_coverage_test:output_hdf5
  Test #27: regression_coverage_test:calculate_pi
  Test #29: regression_coverage_test:advection_convergence
  Test #31: regression_coverage_test:advection_performance

$ ctest -N -LE "coverage"
Test project /gpfs/alpine/ast146/world-shared/build-cuda
  Test  #6: unit_tests:Physical constants
  Test #11: unit_tests:Overlapping SpaceInstances
  Test #17: unit_tests:Time simple stencil operations
  Test #18: unit_tests:Check registry pressure
  Test #19: unit_tests:Check many arrays
  Test #23: performance_tests:Catch2 Container Iterator Performance
  Test #24: regression_test:output_hdf5
  Test #26: regression_test:calculate_pi
  Test #28: regression_test:advection_convergence
  Test #30: regression_test:advection_performance

Total Tests: 10

```
This is unexpected as in the case of code coverage turned off I had expected all tests to not carry the coverage label.

@JoshuaSBrown is this what was intended to happen or is this potentially a bug
In the latter case we may want to add `if`s to the `cmake/TestSetup.cmake` so that coverage tests are not added if they're turned off.

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
## PR Summary

This fixes the problem raised by @pgrete where code coverage tests were being run when the CODE COVERAGE flag was disabled. 

If you follow the commits, there are two solutions, the first makes use of cmake config variables that are inserted into the .cpp files. I am not a fan of this solution because it feel clunky, and it means treating all the unit tests files as config files.

The second solution is much more elegant, but requires bumping the cmake version to 3.12. This one does not require changes to any of the cpp source files. Instead it parses the tests after Catch2 has created them and appends a coverage label to any tests that does not contain the performance label. 

https://github.com/lanl/parthenon/issues/254

## PR Checklist

- [x] Code passes cpplint
- [x] Code is formatted

### CODE REVIEW

**Comment by pgrete** [comment]
Did you confirm that this fully addresses the issue?
I'm still getting that label (`coverage`) based split between tests.

**Comment by JoshuaSBrown** [comment]
> Did you confirm that this fully addresses the issue?
> I'm still getting that label (`coverage`) based split between tests.

Without the coverage flag I get this:
```
[ppc64le] [joshuasbrown@cn2023 build]$ ctest -R coverage
Test project /home/joshuasbrown/Software/parthenon_x86_64/parthenon/build
No tests were found!!!
```

**Comment by JoshuaSBrown** [comment]
Ok, I see what you are getting at, it will be hard to remove the labels from the unit tests because they are added in the actual cpp files not in cmake. The way Catch2 is setup they the labels are parsed in through the TEST_CASE macro in the cpp files, I can see if the coverage label can instead be added through cmake.

**Comment by JoshuaSBrown** [comment]
This works, I had hoped there was a better way to do this, the problem is that the only way to add labels that I can see to catch2 tests is within the .cpp files, I had hoped to use compile definitions but because the catch2 macros that resolve the labels are also handled by the preprocessor the labels are not correctly resolved by doing this. I'm still open to better ways of doing this.

**Review by Yurlungur — APPROVED** [review]
I don't love this solution, but I don't see a better one. If it fixes things, I say we merge it in and don't nitpick. Thanks for digging in and finding a fix.

One potential solution, which I shudder to consider, would be using a text processing like `sed` or `awk`. I do not advocate it though---I think the `cmake` solution is better for us.

**Comment by Yurlungur** [comment]
@JoshuaSBrown looks like formatting check is failing. A `make format`  or a call to `par-hermes` will probably fix it.

**Comment by JoshuaSBrown** [comment]
@par-hermes format

**Comment by JoshuaSBrown** [comment]
> I don't love this solution, but I don't see a better one. If it fixes things, I say we merge it in and don't nitpick. Thanks for digging in and finding a fix.
> 
> One potential solution, which I shudder to consider, would be using a text processing like `sed` or `awk`. I do not advocate it though---I think the `cmake` solution is better for us.

I have now improved it, I like this implementation much more too.

**Review by Yurlungur — APPROVED** [review]
> I have now improved it, I like this implementation much more too.

Aha! Yes, this is a massive improvement. Does any documentation need to change? On a first pass, I think no.

**Comment by JoshuaSBrown** [comment]
> > I have now improved it, I like this implementation much more too.
> 
> Aha! Yes, this is a massive improvement. Does any documentation need to change? On a first pass, I think no.

Thanks! And yes, I need to update some documentation, I want to make sure no one has issues with bumping the cmake version though before proceeding. If the Athena++ guys are ok with this the cmake version used in the ci will also need to be updated.

**Comment by pgrete** [comment]
I agree. This approach looks way cleaner (and a first step towards cleaning up the testing infrastructure).
Feel free to bump the cmake version in both CI files. If I didn't miss anything using `3.12.4` should work in both cases out of the box.

File: CMakeLists.txt [hunk 1]
```diff
@@ -67,7 +67,10 @@ endif()
 
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
+include(cmake/CodeCov.cmake)
+set(CATCH2_COVERAGE_TAG "")
 if(CODE_COVERAGE)
+  set(CATCH2_COVERAGE_TAG "[coverage]")
```
Comment (AndrewGaspar) [inline CMakeLists.txt]: Is this still used?

**Comment by JoshuaSBrown** [comment]
@AndrewGaspar Do I have your approval to merge this?

**Review by AndrewGaspar — APPROVED** [review]

**Pull request merged**